### ISSUE TITLE
ExecuteContext.exception() should accept null exceptions and translate them to an unspecified DataAccessException

### ISSUE THREAD

**Issue by bwajtr** [jOOQ/jOOQ#11304]
### Expected behavior
To be honest, I'm not sure what should be the correct behaviour, but JooqExceptionTranslator should not fail with NPE. The org.springframework.jdbc.core,.JdbcTemplate implementation solves this problem by throwing org.springframework.jdbc.UncategorizedSQLException when the exception translation returns null. So I think that the JooqExceptionTranslator should behave the same...

### Actual behavior
- The DB error is caught by org.springframework.boot.autoconfigure.jooq.JooqExceptionTranslator, which in turn calls SQLErrorCodeSQLExceptionTranslator to translate the error. However, in 5.3.0  version of spring-jdbc the contract of SQLErrorCodeSQLExceptionTranslator#translate() changed from @NotNull to @Nullable (see here https://github.com/spring-projects/spring-framework/commit/e9cded560d6ecb73aa5bfca57e06b09fa0013294#diff-c6aa1c6a4b11e8a722808e945c4c5b6ef471c42871c7ce830554dde81ad7f2c2L68). In the case of the custom errors (or any non-standard error), the SQLErrorCodeSQLExceptionTranslator now returns null (instead of UncategorizedSQLException) and JooqExceptionTranslator fails with NullPointerException, because it does not expect null.
 
```
java.lang.NullPointerException
	at org.jooq.impl.DefaultExecuteContext.exception(DefaultExecuteContext.java:737)
	at org.springframework.boot.autoconfigure.jooq.JooqExceptionTranslator.handle(JooqExceptionTranslator.java:83)
	at org.springframework.boot.autoconfigure.jooq.JooqExceptionTranslator.exception(JooqExceptionTranslator.java:55)
	at org.jooq.impl.ExecuteListeners.exception(ExecuteListeners.java:274)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:390)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:333)
	at org.jooq.impl.AbstractResultQuery.fetchInto(AbstractResultQuery.java:1545)
	at org.jooq.impl.SelectImpl.fetchInto(SelectImpl.java:3936)
```

### Steps to reproduce the problem

- Have any PostgreSQL database
- Use jOOQ in Spring Environment (reproduced with latest Spring Boot 2.4.2, where spring-jdbc was in version 5.3.3)
- Create a function/procedure in PostgreSQL and throw an error in it  (use RAISE EXCEPTION):
    ```sql
    CREATE OR REPLACE PROCEDURE error_test()
    AS $$
    BEGIN
        RAISE EXCEPTION 'This is user defined exception';
    END;
    $$
    LANGUAGE plpgsql;
    ``` 
- Generate jOOQ code for this database
- Call the function it from java code

### Workaround
- We defined our own exception translator by implementing SQLExceptionTranslator interface and then using it by calling 
    ```java
    CustomSQLExceptionTranslatorRegistry.getInstance().registerTranslator("PostgreSQL", 
      new CustomPostgreSQLExceptionTranslator());
    ``` 
- Therefore we made sure that in our usecase the error translation never returns null

### Versions

- jOOQ: 3.14.4
- Java: AdoptOpenJDK jdk-11.0.10+9_openj9-0.24.0
- Database (include vendor): PostgreSQL 11.9
- OS: Windows 10
- JDBC Driver (include name if inofficial driver): org.postgresql:postgresql:42.2.18

**Comment by lukaseder** [comment]
Thanks for your report.

That's quite the incompatible change in Spring, and a bit surprising, too! I'll create an issue in Spring to see what the way forward will be here. Of course, we should avoid the NPE in jOOQ, because it is caused by an unrelated piece of code. The `ExecuteContext.sqlException()` Javadoc reads:

> If null is being passed, jOOQ will internally translate the"unavailable" exception to an unspecified DataAccessException.

So, the same should be done to `ExecuteContext.exception()`. This means you'll still have to register your own translator if you want a specific exception

**Comment by lukaseder** [comment]
The rationale behind the change in Spring is here: https://github.com/spring-projects/spring-framework/issues/24064

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.6 (#11306).

Thanks again for your report and for your time saving research, which is greatly appreciated!

**Comment by bwajtr** [comment]
You are welcome... Thanks for lightning-fast fix :)
